### PR TITLE
Gui/load and show with complete frames

### PR DESCRIPTION
**Pull request by DvoraNovogrotzki**
This branch introduces several improvements and new functionalities in the GUI's handling of loading and displaying frames. The main goal of this update is to ensure the correct and efficient rendering of frames, while also enhancing the overall user experience in terms of performance and visual feedback.

Key Updates:
Loading Complete Frames:

Improved the frame loading logic to ensure that entire frames are loaded and processed correctly before being displayed.
This guarantees that users see a fully constructed frame without any partial updates or missing elements, ensuring a smoother and more cohesive visual experience.
Frame Synchronization:

Enhanced synchronization mechanisms to make sure the GUI updates happen seamlessly, reducing any delays or frame drops during the rendering process.
Ensured that frame data is validated and complete before rendering, avoiding corrupted visuals.
Efficient Memory Management:

Optimized memory usage by ensuring proper handling of frame data, including efficient disposal of old frames and resources.
Incorporated dynamic memory allocation strategies to handle varying frame sizes while minimizing performance overhead.
User Feedback and Loading Indicators:

Added visual indicators to show when frames are loading, providing clear feedback to users during any latency or delay in frame generation.
Improved responsiveness by ensuring that GUI elements remain interactive even during intensive frame processing.
Bug Fixes:

Fixed issues related to frame alignment and resizing when rendering in different window sizes or aspect ratios.
Addressed edge cases where incomplete frames were displayed due to asynchronous loading conflicts.
Code Refactoring:

Cleaned up the code to improve readability and maintainability, especially around the frame loading and rendering functions.
Applied best practices for error handling to ensure robustness, especially in scenarios where frame data is not available or corrupt.

### CODE REVIEW

File: gui/src/draggable_square.cpp [hunk 1]
```diff
@@ -165,15 +166,15 @@ void DraggableSquare::mouseMoveEvent(QMouseEvent *event)
     }
 
     QPoint newPos = mapToParent(event->pos() - dragStartPosition);
-    newPos.setX(qMax(0, qMin(newPos.x(), parentWidget()->width() - width())));
-    newPos.setY(qMax(0, qMin(newPos.y(), parentWidget()->height() - height())));
+    QRect parentRect = parentWidget()->rect();
+
+    // חישוב מיקום חדש, בהסתמך על הגבולות של ההורה
```
Comment (ChedvaErblich) [inline gui/src/draggable_square.cpp]: Write your comments in English.

File: gui/src/frames.cpp [hunk 2]
```diff
@@ -49,27 +50,32 @@ void Frames::paintEvent(QPaintEvent *event)
...
+                QPen pen(QColor(frame.color), frame.thickness);
+                pen.setJoinStyle(Qt::RoundJoin);
+                painter.setPen(pen);
 
                 auto square1 = logHandler.getProcessSquares()[log.srcId];
                 auto square2 = logHandler.getProcessSquares()[log.dstId];
 
-                QRect rect1(square1->getDragStartPosition().x(),
-                            square1->getDragStartPosition().y(),
+                QRect rect1(square1->pos().x() - 8, square1->pos().y() - 8,
```
Comment (ChedvaErblich) [inline gui/src/frames.cpp]: If possible, it's better to avoid using hardcoded numbers and define them instead.

File: gui/src/frames.cpp [hunk 3]
```diff
@@ -49,27 +50,32 @@ void Frames::paintEvent(QPaintEvent *event)
...
                             square1->width(), square1->height());
-                QRect rect2(square2->getDragStartPosition().x(),
-                            square2->getDragStartPosition().y(),
+                QRect rect2(square2->pos().x() - 8, square2->pos().y() - 8,
                             square2->width(), square2->height());
-                painter.drawRect(rect1);
-                painter.drawRect(rect2);
+
+                painter.drawRoundedRect(rect1, 10, 10);
```
Comment (ChedvaErblich) [inline gui/src/frames.cpp]: If possible, it's better to avoid using hardcoded numbers and define them instead.

File: gui/src/log_handler.cpp [hunk 4]
```diff
@@ -105,23 +109,24 @@ void LogHandler::analyzeLogEntries(QMainWindow *mainWindow,
...
 
-            Process process(id, name, cmakeProject, qemuPlatform);
+            Process *process =
+                new Process(id, name, cmakeProject, qemuPlatform);
             DraggableSquare *square =
                 new DraggableSquare(mainWindow, "", width, height);
-            square->setProcess(&process);
-            square->setDragStartPosition(QPoint(x, y));
+            square->setProcess(process);
+            // square->setDragStartPosition(QPoint(x, y));
```
Comment (ChedvaErblich) [inline gui/src/log_handler.cpp]: Why is this written as a comment? If there's no good reason, delete it.

File: gui/src/simulation_state_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    for (int i = 0; i < squares.size(); i++) {
+        if (squares[i] != nullptr) {
+            data.squares.append(squares[i]);
+            MainWindow::guiLogger.logMessage(
+                logger::LogLevel::DEBUG, "SimulationStateManager",
+                "readSimulationState",
+                "Added process at index " + std::to_string(i));
+        }
+        else
+            MainWindow::guiLogger.logMessage(
```
Comment (ChedvaErblich) [inline gui/src/simulation_state_manager.cpp:42]: What is the severity level of the error? 
Is logging an error message sufficient, or does it require more serious handling?

File: gui/test/test_main_window.cpp [hunk 6]
```diff
@@ -5,7 +5,7 @@ class TestMainWindow : public QObject {
     Q_OBJECT
 
 private slots:
-    void init();  // Setup method, runs before each test
+    void init();     // Setup method, runs before each test
```
Comment (ChedvaErblich) [inline gui/test/test_main_window.cpp:8]: Why?

File: gui/test/test_process_dialog.cpp [hunk 7]
```diff
@@ -43,14 +43,15 @@ private slots:
...
+    void testValidateAndAccept()
     {
         ProcessDialog dialog;
         dialog.setId(123);
         dialog.setName("Test Process");
         dialog.setCMakeProject("TestProject");
         dialog.setQEMUPlatform("x86");
-        QVERIFY(dialog.validateAndAccept() == true); // Assuming true indicates validation success
+        QVERIFY(dialog.validateAndAccept() ==
+                true);  // Assuming true indicates validation success
```
Comment (ChedvaErblich) [inline gui/test/test_process_dialog.cpp]: Place the comment on the line before to avoid splitting the condition, making the code more readable.

File: logger/logger.cpp
Comment (ChedvaErblich) [inline logger/logger.cpp:1]: Don't modify other users' code. If we want to address convention issues in this code, it should be done in a separate commit and coordinated with the code owner.

File: logger/logger.h
Comment (ChedvaErblich) [inline logger/logger.h:1]: Don't modify other users' code. If we want to address convention issues in this code, it should be done in a separate commit and coordinated with the code owner.

File: gui/test/test_process_dialog.cpp (hunk 7 above)
Comment (DvoraNovogrotzki) [reply gui/test/test_process_dialog.cpp]: Done.

File: gui/src/simulation_state_manager.cpp (hunk 5 above)
Comment (DvoraNovogrotzki) [reply gui/src/simulation_state_manager.cpp:42]: This handles an error where the array of squares contains a null pointer, meaning one square has already been deleted. In my opinion, this doesn't require special handling because there's no reason for this to cause an issue.

File: logger/logger.h
Comment (DvoraNovogrotzki) [reply logger/logger.h:1]: I'm sorry I'll not do it again...

File: gui/test/test_main_window.cpp (hunk 6 above)
Comment (DvoraNovogrotzki) [reply gui/test/test_main_window.cpp:8]: The formater made it. I'll not do it again... (to modify other users' code)

File: gui/src/frames.cpp (hunk 3 above)
Comment (DvoraNovogrotzki) [reply gui/src/frames.cpp]: Done.

File: gui/src/log_handler.cpp (hunk 4 above)
Comment (DvoraNovogrotzki) [reply gui/src/log_handler.cpp]: Done.

File: gui/src/frames.cpp (hunk 2 above)
Comment (DvoraNovogrotzki) [reply gui/src/frames.cpp]: Done.

File: gui/src/draggable_square.cpp (hunk 1 above)
Comment (DvoraNovogrotzki) [reply gui/src/draggable_square.cpp]: Done.

File: logger/logger.h
Comment (ChedvaErblich) [reply logger/logger.h:1]: NP 👍
